src/lib/pci_virtio_net_vpnkit.c

```diff
@@ -324,21 +343,60 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
 		init_reply.magic[4],
 		init_reply.version, (int)sizeof(init_reply.commit), init_reply.commit);
 
-	struct msg_ethernet cmd_ethernet = {
-		.command = CMD_ETHERNET,
-	};
-	memcpy(cmd_ethernet.uuid, uuid, sizeof(cmd_ethernet.uuid));
-
-	if (really_write(fd, (uint8_t*)&cmd_ethernet, sizeof(cmd_ethernet)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to write ethernet cmd\n");
-		return -1;
+	if (init_reply.version != init_msg.version) { 
+		fprintf(stderr, "virtio-net-vpnkit: protocol version mismatch: version %d requested, got version %d.\n", 
+			init_msg.version,
+			init_reply.version);
 	}
 
-	if (really_read(fd, (uint8_t*)vif, sizeof(*vif)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to read vif info\n");
+	if (preferred_ipv4 == NULL) { // ethernet
```

Comment:
TBH I'm not really sure what this comment is supposed to tell me, regardless of where it is placed.

I'd prefer if the if/else was inverted so it was `preferred_ipv4 != NULL { do stuff using preferred_ipv4 } else { do the other thing }`. With that I also think there would be no need for a comment for the `preferred_ipv4 != NULL` case since it would be self explanatory, the `else` case would have a `/* If no preferred ipv4 address configured then we «explain the fallback» */`

Reply:
Hm, yes - I agree that would be more readable, I'll swap them around and remove the pointer as suggested below